### PR TITLE
Lesson05-07: Implement character device driver with procfs/sysfs interfaces

### PR DESCRIPTION
**Pull request by yorkened**
# Summary
This is my implementation of character device driver for text messaging between users. Such steps were followed during development:
- Created new Linux kernel module;
- Implemented read/write functions from file_operations struct;
- Provided buffer increase as a module parameter;
- Provided work only with ASCII strings;
- Implemented procfs interface to get information about current buffer size and capacity;
- Implemented sysfs interface for buffer clearance.
## Device driver testing scenario
1. Check if module is loaded and relevant device and interfaces exist.
2. Print information about initial buffer state from procfs inerface.
3. Write data to character device.
4. Check if info about written data available from procfs interface.
5. Read all data from character device.
6. Check if buffer is clear now.
7. Write data to device again and clear buffer with sysfs interface.
8. Try to write data blocks to character device from /dev/urandom to check if driver deny writing of non-ASCII strings.

### CODE REVIEW

**Review by galkinletter — CHANGES_REQUESTED** [review]
Makefile missing

**Comment by yorkened** [comment]
@galkinletter, but there is no mentioning in task that makefile source is needed

**Comment by galkinletter** [comment]
For all the following tasks, it should be the default

**Comment by yorkened** [comment]
@galkinletter, so I just need to upload my makefile used during building of module?

**Comment by yorkened** [comment]
Done. Makefile was added
@galkinletter please, review my request

File: 02-KernelModules_CharacterDevice/src/Makefile [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+# Char dev Linux kernel module
+#
+
+obj-m := intercom.o
+
+all:
+	$(MAKE) -C $(BUILD_KERNEL) M=$(CURDIR) modules
+
+clean:
+	$(MAKE) -C $(BUILD_KERNEL) M=$(CURDIR) clean
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/src/Makefile:11]: Need Fix - add empty line at the end of file
```suggestion
	$(MAKE) -C $(BUILD_KERNEL) M=$(CURDIR) clean
	
```

File: 02-KernelModules_CharacterDevice/src/intercom.c [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+	pr_info("Exiting intercom\n");
+}
+
+module_init(intercom_init);
+module_exit(intercom_exit);
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Nikita Kvochka <yorkened@gmail.com>");
+MODULE_DESCRIPTION("Simple char dev for text messaging between users");
+MODULE_VERSION("0.1");
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/src/intercom.c:275]: add empty line at the end of file
```suggestion
MODULE_VERSION("0.1");

```

File: 02-KernelModules_CharacterDevice/test/devtest.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+cat $PROCFS
+
+printf "\n### Testing sysfs interface ###\n"
+echo "Testing clearance of buffer with sysfs interface" > $DEVICE
+cat $PROCFS
+cat $SYSFS
+cat $PROCFS
+
+printf "\n### Testing for ASCII check ###\n"
+dd if=/dev/urandom of=$DEVICE bs=512 count=4
```
Comment (galkinletter) [inline 02-KernelModules_CharacterDevice/test/devtest.sh:33]: add empty line at the end of file

**Review by galkinletter — CHANGES_REQUESTED** [review]
Please fix empty line at the end of file

**Comment by galkinletter** [comment]
Thank you for your attention to error handling.

**Comment by yorkened** [comment]
@galkinletter sorry, but my sources already have newlines at the end of files. Github just trim them all for good-looking web view. On the other hand, if it was no newlines at the end of files, you will see following pop-up notification. And if you, for example, select "view file" for my makefile, you will see that there are 12 lines in file, but github shows only 11 of them.

изображение

изображение

File: 02-KernelModules_CharacterDevice/src/intercom.c [hunk 4]
```diff
@@ -0,0 +1,274 @@
...
+	.write = dev_write
+};
+
+static struct file_operations procfs_fops = {
+	.read = procfs_read,
+};
+
+static int intercom_uevent(struct device *dev, struct kobj_uevent_env *env)
+{
+	add_uevent_var(env, "DEVMODE=%#o", 0666);
```
Comment (oleg-h2o) [inline 02-KernelModules_CharacterDevice/src/intercom.c]: if(add_uevent....)
    return -ENOMEM;

**Review by oleg-h2o — CHANGES_REQUESTED** [review]
Now it's almost perfect, but a little fix is required

**Comment by yorkened** [comment]
@oleg-h2o, fix is ready, please, make review.

**Pull request merged**